lib/active_storage/service/s3_service.rb

```diff
@@ -4,8 +4,14 @@
 class ActiveStorage::Service::S3Service < ActiveStorage::Service
   attr_reader :client, :bucket
 
-  def initialize(access_key_id:, secret_access_key:, region:, bucket:)
-    @client = Aws::S3::Resource.new(access_key_id: access_key_id, secret_access_key: secret_access_key, region: region)
+  def initialize(access_key_id:, secret_access_key:, region:, bucket:, endpoint: nil)
+    params = {
+      access_key_id:     access_key_id,
+      secret_access_key: secret_access_key,
+      region:            region
+    }
+    params.merge!(endpoint ? {endpoint: endpoint, force_path_style: true} : {})
```

Comment:
Okay. In that case, I think I’d rather repeat the call to `Aws::S3::Resource.new` than make building up an options hash the focus of the method:

```ruby
def initialize(access_key_id:, secret_access_key:, region:, bucket:, endpoint: nil)
  @client = if endpoint
    Aws::S3::Resource.new(
      access_key_id: access_key_id,
      secret_access_key: secret_access_key,
      region: region,
      bucket: bucket
    )
  else
    Aws::S3::Resource.new(
      access_key_id: access_key_id,
      secret_access_key: secret_access_key,
      region: region,
      bucket: bucket,
      endpoint: endpoint
    )
  end

  @bucket = @client.bucket(bucket)
end
```

Once you make that change, I’ll merge.